service/controller/resource/appversionlabel/create.go

```diff
@@ -0,0 +1,101 @@
+package appversionlabel
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/apiextensions/v3/pkg/label"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/giantswarm/cluster-operator/v3/pkg/project"
+	"github.com/giantswarm/cluster-operator/v3/service/controller/key"
+	"github.com/giantswarm/cluster-operator/v3/service/internal/releaseversion"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// The app custom resources are deleted when the namespace is deleted.
+	if key.IsDeleted(&cr) {
+		r.logger.Debugf(ctx, "not updating app version labels for tenant cluster %#q", key.ClusterID(&cr))
+		r.logger.Debugf(ctx, "canceling resource")
+		return nil
+	}
+
+	var apps []*v1alpha1.App
+	{
+		r.logger.Debugf(ctx, "finding optional apps for tenant cluster %#q", key.ClusterID(&cr))
+
+		o := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s!=%s", label.ManagedBy, project.Name()),
+		}
+		list, err := r.g8sClient.ApplicationV1alpha1().Apps(key.ClusterID(&cr)).List(ctx, o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, item := range list.Items {
+			apps = append(apps, item.DeepCopy())
+		}
+
+		r.logger.Debugf(ctx, "found %d optional apps for tenant cluster %#q", len(apps), key.ClusterID(&cr))
+	}
+
+	var updatedAppCount int
```

Comment:
We could move the variable inside the other scopes because it is actually not used outside here.